### ISSUE TITLE
New Doc Proposal: Converting to a Composer Managed Workflow

### ISSUE THREAD

**Issue by dustinleblanc** [pantheon-systems/documentation#3053]
## Issue Description

Drupal 8 users are increasingly stumbling into the land of Composer infused limbo when they require their first module via Composer and then run into merge conflicts during the standard upstream update process.

The standard answer for this is to convert to a Composer managed workflow (a la build-tools) but the process of converting a site in place is neither straightforward or pain free for many customers. Creating a brand new site using build tools is one option, but has it's own drawbacks (especially related to site billing/config).

## Suggested Resolution

Document a standard set of steps to convert an existing site to a Composer managed workflow so that we can give folks a standard explanation of the journey ahead when they wind up in this unfortunate dependency management limbo.

@greg-1-anderson, @stevector - You guys would be great contributors I think.

**Comment by greg-1-anderson** [comment]
What we usually recommend is:

- Create a new site with `terminus build:project:create`
- Use `composer require` to add any contrib modules previously added to the current site
- Commit any custom modules and other custom code / assets to the new repository
- Copy the database from the new site to the current site

This has worked well for a number of customers. The largest pain points I can see here would be for folks who may not fully understand what code is in their site (esp. for inherited sites). The [composer_generate](https://www.drupal.org/project/composer_generate) project can help here for modules managed on drupal.org, but users are still on the hook to identify all of their custom code.

@dustinleblanc Do you have any insight into any other pain points here?

**Comment by dustinleblanc** [comment]
Creation of a new site is definitely a good idea, and the simplest path, but I know multiple customers have specified a preference to not have to do that, which I understand. I've personally done the 'in place' method and successfully done so without requiring CI setups either.

Those are the two biggest barriers that I've heard, CI requirement and new site requirement. Being able to just 'replace in place' and develop a 'composer locally and just commit everything' method would probably help aid in adoption / reduce friction.

I wonder if it would be crazy to create a Terminus plugin for swapping a site on a local machine? 

The steps I'd see are

1. New git branch
2. sweep code into a temp location
3. Drop in example-drops8-composer
4. pull in composer.json customizations/modules/libraries
5. Try to find Composer equivelants for contrib projects
6. Identify custom code
7. copy custom code back in
8. Commit the changes

Those are the steps I think whether a human or computer tries to do them

**Comment by greg-1-anderson** [comment]
If the use-case is to set up without a CI workflow, then doing it in place would not be super hard. You'd just have to accept thereafter that you'd have to do all of your composer operations locally, and your repo would contain all of the assets for your site (e.g. vendor), just as it does prior to the conversion. The tricky part is 6, identify custom code.

The problem with in-place conversion using the GitHub/Circle pull request workflow is that the script would need to figure out exactly how your Drupal site was set up, and would have to create build steps that did the right thing for whatever decision the site builder happened to make. I'm not sure how hard that part would be; if you don't want automation or a separate GitHub repository, then perhaps what you suggest would be workable. However, the problem here is that we no longer support the manual installation method for example-drops8-composer.

In addition to the steps you mention, it would also be necessary to:

1. Convert to a relocated docroot.
2. Switch upstreams to the correct "empty" upstream.

**Comment by dustinleblanc** [comment]
1. would be part of my step 3 (since it contains the edits to the pantheon.yml)
2. Yep, very true, they could do this after confirming everything works

Identifying the custom code is definitely a tricky part. If they follow sane conventions, then its not so bad, but we can't guarantee that unfortunately

**Comment by clayball** [comment]
How does a customized upstream fit into this process?

**Comment by greg-1-anderson** [comment]
@clayball In the [Pantheon custom upstream process](https://pantheon.io/docs/custom-upstream/), you can add new features to your upstream repository, and pull those improvements into any site created from that upstream from the Pantheon dashboard.

In contrast, the Composer workflow assumes that all sites will be updated only via Composer. If you have any shared code that you'd like to be able to share across multiple sites, you could package it as a module or a plugin, and then `composer require` that project from each site that uses it. Then, if you add new features to your module / plugin, you would use `composer update` (on a local copy of your site) to pull in the new features.

**Comment by clayball** [comment]
@greg-1-anderson Yes, that makes sense. However, we have a different workflow. We've recently created a custom upstream based upon [https://github.com/pantheon-systems/example-drops-8-composer](https://github.com/pantheon-systems/example-drops-8-composer)

Our upstream is now composer based. So, we do `composer [require|update]` in our upstream repo and then our sites merge in the changes/updates.

The one issue we've run into is that now sites are not able to do composer stuff because the upstream is doing composer stuff. So, if, for example, a site wants to test a new module, one option might be to create a multi dev, run `composer require`, test. If testing goes well and the module passes security review then we do `composer require` in the upstream repo and all sites will have the option of using the new module.

Does that make sense?
Is the workflow I mentioned documented anywhere?
Does it fit into this 'New Doc Proposal'?

Thanks,
Clay

**Comment by greg-1-anderson** [comment]
@clayball Using a derivative of example-drops-8-composer as an upstream as you are doing is supported. To add a module to the downstream sites, add said module to your upstream, and then do a dashboard update on each site. Your new module will appear in every site.

Adding a module directly to one of the sites that uses your upstream without adding it to the upstream is not supported. If you do this, you will get merge conflicts when you attempt to do a dashboard update to a newer version of your upstream.

**Comment by clayball** [comment]
@greg-1-anderson Excellent!

I should have noted that the multi-dev I mentioned for testing a new module would never be merged back into dev due to, as you mentioned, the merge conflicts that would occur :)

Thanks!
